content/blog/2025-07-14-user-defined-parquet-indexes.md

```diff
@@ -0,0 +1,545 @@
+
+2. Calls `DistinctIndex::new_from_reader` to read the index from the file at that offset.
+
+`DistinctIndex::new_from_reader` actually reads the index as shown below:
+
+```rust
+ /// Read the distinct values index from a reader at the given offset and length
+ pub fn new_from_reader<R: Read + Seek>(mut reader: R, offset: u64) -> Result<DistinctIndex> {
+     reader.seek(SeekFrom::Start(offset))?;
+
+     let mut magic_buf = [0u8; 4];
+     reader.read_exact(&mut magic_buf)?;
+     if magic_buf != INDEX_MAGIC {
+         return exec_err!("Invalid index magic number at offset {offset}");
+     }
+
+     let mut len_buf = [0u8; 8];
+     reader.read_exact(&mut len_buf)?;
+     let stored_len = u64::from_le_bytes(len_buf) as usize;
+
+     let mut index_buf = vec![0u8; stored_len];
+     reader.read_exact(&mut index_buf)?;
+
+     let Ok(s) = String::from_utf8(index_buf) else {
+         return exec_err!("Invalid UTF-8 in index data");
+     };
+
+     Ok(Self {
+         inner: s.lines().map(|s| s.to_string()).collect(),
+     })
+ }
+```
+
+This code:
+
+1. Seeks to the offset of the index in the file.
+
+2. Reads the magic bytes and checks they match `IDX1`.
+
+3. Reads the length of the index and allocates a buffer.
+
+4. Reads the index bytes, converts them to a `String`, and splits into lines to populate the `HashSet<String>`.
+
+
+### Extending DataFusion’s `TableProvider`
+
+---
+
+To use the distinct index for file-level pruning, extend DataFusion's `TableProvider` to read the index and apply it during query execution:
+
+```rust
+impl TableProvider for DistinctIndexTable {
+    /* ... */
+
+    /// Prune files before reading: only keep files whose distinct set
+    /// contains the filter value
+    async fn scan(
+        &self,
+        _ctx: &dyn Session,
+        _proj: Option<&Vec<usize>>,
+        filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        // This example only handles filters of the form
+        // `category = 'X'` where X is a string literal
+        //
+        // You can use `PruningPredicate` for much more general range and
+        // equality analysis or write your own custom logic.
+        let mut target: Option<&str> = None;
+
+        if filters.len() == 1 {
+            if let Expr::BinaryExpr(expr) = &filters[0] {
+                if expr.op == Operator::Eq {
+                    if let (
+                        Expr::Column(c),
+                        Expr::Literal(ScalarValue::Utf8(Some(v)), _),
+                    ) = (&*expr.left, &*expr.right)
+                    {
+                        if c.name == "category" {
+                            println!("Filtering for category: {v}");
+                            target = Some(v);
+                        }
+                    }
+                }
+            }
+        }
+        // Determine which files to scan
+        let files_to_scan: Vec<_> = self
+            .files_and_index
```

Comment:
Might be worth calling out that `fn files_and_index` is left as an exercise for the reader... I spent a bit looking around for it.

Reply:
added in  3ccfeb3